File: python/src/runapi/omnihuman/resources/human_identification.py
"""OmniHuman human-identification resource."""

from __future__ import annotations

from typing import Any, Dict, Optional

from runapi.core import Resource, RequestOptions

from ..contract_gen import CONTRACT
from ..types import (
    CompletedHumanIdentificationResponse,
    HumanIdentificationResponse,
)


class HumanIdentification(Resource):
    """Identify human regions in a source image."""

    ENDPOINT = "/api/v1/omnihuman/human_identification"

    RESPONSE_CLASS = HumanIdentificationResponse
    COMPLETED_RESPONSE_CLASS = CompletedHumanIdentificationResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a task and poll until it completes."""
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a human-identification task and return immediately with an ``id``."""
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        """Fetch the current status of a human-identification task."""
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)

    def _validate_params(self, params: Dict[str, Any]) -> None:
        self._validate_contract(CONTRACT["human-identification"], params)

File: ruby/lib/runapi/omnihuman/resources/human_identification.rb
# frozen_string_literal: true

module RunApi
  module Omnihuman
    module Resources
      # Identifies human regions in a source image before video generation.
      class HumanIdentification
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/omnihuman/human_identification"
        RESPONSE_CLASS = Types::HumanIdentificationResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedHumanIdentificationResponse

        def initialize(http)
          @http = http
        end

        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        def create(options: nil, **params)
          params = compact_params(params)
          validate_contract!(CONTRACT["human-identification"], params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end
      end
    end
  end
end
